server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.repositories;
+
+import org.apache.lucene.index.IndexCommit;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.snapshots.IndexShardSnapshotStatus;
+import org.elasticsearch.index.store.Store;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Manages the {@link IndexCommit} associated with the shard snapshot as well as its {@link IndexShardSnapshotStatus} instance.
+ */
+public class ShardSnapshotContext {
```

Comment:
At first sight I liked this new class, but after iterating over this PR few times I like it a bit less. What I don't like here is that it adds another place where the `IndexShardSnapshotStatus` is modified (and kind of hide these modifications), and I also feel like it adds extra complexity.

Instead, I think we could wrap the IndexCommit into a AbstractRefCounted, increment it before triggering a file snapshot, decrement it once the snapshot of the file is done and let the IndexCommit be released itself once all file snapshots are done.

Reply:
> Instead, I think we could wrap the IndexCommit into a AbstractRefCounted, increment it before triggering a file snapshot, decrement it once the snapshot of the file is done and let the IndexCommit be released itself once all file snapshots are done.

That was my plan A at first. But then I started not liking that so much because:
* It makes the signature `org.elasticsearch.repositories.blobstore.BlobStoreRepository#snapshotShard` even more complex than it already is
* While it makes the state transitions of `IndexShardSnapshotStatus` a little more obvious now, it hides the connection between that ref counted instance, the index commit and the status doesn't it?

... Maybe it would be better to just flatten all this logic into `IndexShardSnapshotStatus` (or the logic of the shard status tracking into `ShardSnapshotContext`)? :)

The other thing about me adding this was that we will need to keep track of more state for each shard upload when fixing/optimizing the situation of unchanged shards in https://github.com/elastic/elasticsearch/issues/45736 ... my thinking was that if we change the signature like it was done here, we could do without yet another subsequent signature change across all repository implementations when dealing with that and simply add the necessary logic to convey an unchanged shard and its generation via `ShardSnapshotContext`?

Reply:
It looks to me that we are trying to address different issues in this PR:
a) more efficient way to snapshot shard files
b) better visibility and better tracking of shard status and resources releases
c) prepare the work for #45736

I think this PR should be just about point a). I'm not saying that b) and c) are not valid and I'm pretty sure they deserve their own PRs :) but for now I don't think they should be an argument to introduce more complexity for what we want to solve here.

What we could do (sorry again if I wasn't clear before) is to acquire an index commit reference in  `SnapshotShardsService.snapshot()` like it was before, and ensure that the reference is released when the shard snapshot listener gets called. Then the listener is passed down to the `BlobStoreRepository#snapshotShard()`  which makes sure to call it through the group files listener. This way we don't touch how status is tracked and we don't introduce another complex object (which again, could be the subject of a dedicated PR). WDYT?

Reply:
@tlrx I think my main point above all the others was `c)` (sorry for not making that clear) and the fact that we would have to change the signature of `snapshotShard` twice (especially when those versions of the signature may be spread out across releases).

I think other than that, I just liked this approach because it kept the logic a little less spread out. But if the above point can't warm your heart to this change I can probably do what you're asking for pretty easily :)

Reply:
Done :) Now we're doing all the resource handling in the callback and are still creating the commit ref in the snapshot shards service like we did before.
Thanks for pushing on this ... turns out this massively simplified the change for the time being (200+ LoC changes saved :)) ... not sure why I did it so complicated the first time around.